### PR TITLE
Migrate Model Indexes to RTK

### LINKED ISSUE #42503: use RTK Query for model-indexes API

### PR DESCRIPTION
**Pull request by npfitz**
Closes https://github.com/metabase/metabase/issues/42503
### Description
Migrating Model Index API to RTK Query. There shouldn't be any differences to the end user experience

### How to verify
Green CI

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx [hunk 1]
```diff
@@ -7,7 +7,8 @@ import { usePrevious } from "react-use";
 import { t } from "ttag";
 import _ from "underscore";
 
-import { useModelIndexesListQuery } from "metabase/common/hooks";
+// import { useModelIndexesListQuery } from "metabase/common/hooks";
```
Comment (rafpaf) [inline frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx:10]: needed?

**Review by rafpaf — APPROVED** [review]

**Review by rafpaf — APPROVED** [review]

File: frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx (hunk 1 above)
Comment (npfitz) [reply frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx:10]: Whoops 😬

File: frontend/src/metabase/entities/model-indexes/actions.ts [hunk 2]
```diff
@@ -1,34 +1,45 @@
...
 
     if (fieldsWithIndexFlags.length === 0) {
       return;
     }
 
     const existingIndexes: ModelIndex[] =
-      ModelIndexes.selectors.getList(getState(), {
-        entityQuery: { model_id: model.id() },
-      }) ?? [];
+      listModelIndexes.select({ model_id: model.id() })(getState()).data ?? [];
```
Comment (sloansparger) [inline frontend/src/metabase/entities/model-indexes/actions.ts:42]: These selectors are a bit of a footgun. They select the current value of the cache. However, the cache is cleared when there are no active subscribers to that endpoint + input. It's safer to fetch the data again through RTK, which will select from the cache for you if there are subscribers, but also refetch for you if there is not.

File: frontend/src/metabase/api/model-index.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        ...invalidateTags(error, [listTag("model-index")]),
+      ],
+    }),
+  }),
+});
+
+export const {
+  useCreateModelIndexMutation,
+  useListModelIndexesQuery,
+  endpoints: { createModelIndex, listModelIndexes, deleteModelIndex },
```
Comment (sloansparger) [inline frontend/src/metabase/api/model-index.ts:47]: TIL you can do a nested destructure on an export

**Review by sloansparger — CHANGES_REQUESTED** [review]

File: frontend/src/metabase/entities/model-indexes/actions.ts (hunk 2 above)
Comment (npfitz) [reply frontend/src/metabase/entities/model-indexes/actions.ts]: I updated this line to dispatch an `initiate` instead, though I'm not really sure what should happen on a failure :/. From what I can see on the network tab, it's still grabbing from the store though

File: frontend/src/metabase/api/tags/utils.ts [hunk 4]
```diff
@@ -190,6 +191,21 @@ export function provideCollectionTags(
...
+): TagDescription<TagType>[] {
+  return [idTag("model-index", modelIndex.id)];
+}
+
+export function provideModelIndexListTags(
+  modelIndexs: ModelIndex[],
+): TagDescription<TagType>[] {
+  return [
+    listTag("model-index"),
+    ...modelIndexs.flatMap(modelIndex => provideModelIndexTags(modelIndex)),
```
Comment (sloansparger) [inline frontend/src/metabase/api/tags/utils.ts:205]: `modelIndexs` => `modelIndexes`
sorry, didn't notice this before 😓

File: frontend/src/metabase/entities/model-indexes/actions.ts [hunk 5]
```diff
@@ -1,34 +1,46 @@
...
       return;
     }
 
-    const existingIndexes: ModelIndex[] =
-      ModelIndexes.selectors.getList(getState(), {
-        entityQuery: { model_id: model.id() },
-      }) ?? [];
+    const { data: existingIndexes = [] } = await (dispatch(
+      listModelIndexes.initiate({ model_id: model.id() }),
+    ) as Promise<{ data: ModelIndex[] }>);
```
Comment (sloansparger) [inline frontend/src/metabase/entities/model-indexes/actions.ts:43]: would it be better to throw if there's a failure? i think we're likely to generate to failed/incorrect requests below if this requests fails.

File: frontend/src/metabase/entities/model-indexes/actions.unit.spec.ts [hunk 6]
```diff
@@ -59,12 +77,11 @@ describe("Entities > model-indexes > actions", () => {
         createMockField(),
       ]);
 
-      const mockDispatch = jest.fn();
-      const mockGetState = jest.fn();
-      await updateModelIndexes(model)(mockDispatch, mockGetState);
+      const { dispatch, getState, dispatchSpy, getStateSpy } = setup();
+      await updateModelIndexes(model)(dispatch, getState);
```
Comment (sloansparger) [inline frontend/src/metabase/entities/model-indexes/actions.unit.spec.ts:81]: weird type issue, looks like it matches the signature from a glance. happy to look at it with you if needed.

**Review by sloansparger — CHANGES_REQUESTED** [review]

**Review by sloansparger — APPROVED** [review]

**Pull request merged**